### PR TITLE
#3 infinities

### PR DESCRIPTION
**Pull request by maxonfjvipon**
Introduced infinity objects

### CODE REVIEW

**Comment by Graur** [comment]
@maxonfjvipon Just some offers, all of them is up to you. See above.

**Comment by maxonfjvipon** [comment]
@Graur Sorry, I don't see, did you left some review or comments? Or what offers are you talking about?

**Comment by Graur** [comment]
@maxonfjvipon I left some comments in your pull request. Can you see them when you scroll up this page?

**Comment by maxonfjvipon** [comment]
@Graur No, can't see it, checked everywhere

**Comment by Graur** [comment]
@maxonfjvipon Ooops, my bad. I forgot to push the submit button, sorry

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  # Negation of $
+  [] > neg
+    positive-infinity > @
+
+  # Sign of $ (-1.0, 0.0, or 1.0)
+  [] > signum
+    -1.0 > @
+
+  # Difference between $ and x
+  [x] > sub
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        x.as-bytes.eq (nan.as-bytes)
+        x.eq 0.0
+      nan
+      if.
+        x.eq ^
+        positive-infinity
+        ^
+
+  # Sum of $ and x
+  [x] > add
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  # Tests that $ > x
+  [x] > greater
+    FALSE > @
+
+  # Tests that $ ≥ x
+  [x] > geq
+    ^.eq x > @
+
+  # Multiplication of $ and x
+  [x] > mul
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      eq. > @
+        x.as-bytes
+        nan.as-bytes
+
+  # Tests that $ > x
+  [x] > greater
+    FALSE > @
+
+  # Tests that $ ≥ x
+  [x] > geq
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+  # Tests that $ ≤ x
+  [x] > leq
+    not > @
+      eq. > @
+        x.as-bytes
+        nan.as-bytes
+
+  # Tests that $ > x
+  [x] > greater
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+  # Tests that $ > x
+  [x] > greater
+    not. > @
+      or.
+        x.as-bytes.eq (nan.as-bytes) # x is nan
+        ^.eq x # x is positive-infinity
+
+  # Tests that $ ≥ x
+  [x] > geq
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  # Tests that $ < x
+  [x] > less
+    FALSE > @
+
+  # Tests that $ ≤ x
+  [x] > leq
+    ^.eq x > @
+
+  # Tests that $ > x
+  [x] > greater
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  # Tests that $ ≠ x
+  [x] > neq
+    (^.eq x).not > @
+
+  # Tests that $ < x
+  [x] > less
+    FALSE > @
+
+  # Tests that $ ≤ x
+  [x] > leq
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  # Negation of $
+  [] > neg
+    negative-infinity > @
+
+  # Sign of $ (-1.0, 0.0, or 1.0)
+  [] > signum
+    1.0 > @
+
+  # Difference between $ and x
+  #[x] > sub
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        ^.eq x # x is positive-infinity
+
+  # Tests that $ ≥ x
+  [x] > geq
+    eq. > @
+      x.as-bytes
+      nan.as-bytes
+
+  # Multiplication of $ and x
+  #[x] > mul
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  #  [] > @
+  #    "You can't manipulate with NaN" > msg
+
+  # Sum of $ and x
+  #[x] > add
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+  # Tests that $ ≥ x
+  [x] > geq
+    not. > @
+      eq.
+        x.as-bytes
+        nan.as-bytes
+
+  # Multiplication of $ and x
+  [x] > mul
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        x.eq 0.0
+      nan
+      if.
+        x.eq negative-infinity
+        negative-infinity
+        ^
+
+
+  # Sum of $ and x
+  [x] > add
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/positive-infinity.eo [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+  # Negation of $
+  [] > neg
+    negative-infinity > @
+
+  # Sign of $ (-1.0, 0.0, or 1.0)
+  [] > signum
+    1.0 > @
+
+  # Difference between $ and x
+  [x] > sub
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/positive-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    eq. > @
+      ^.as-bytes
+      x.as-bytes
+
+  # Tests that $ ≠ x
+  [x] > neq
+    (^.eq x).not > @
+
+  # Tests that $ < x
+  [x] > less
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

File: src/main/eo/org/eolang/math/negative-infinity.eo [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    not. > @
+      or. > @
+        x.as-bytes.eq (nan.as-bytes) # x is nan
+        ^.eq x # x is negative-infinity
+
+  # Tests that $ ≤ x
+  [x] > leq
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply src/main/eo/org/eolang/math/negative-infinity.eo]: Done!

**Comment by Graur** [comment]
@maxonfjvipon Looks great!

**Review by Graur — APPROVED** [review]

**Comment by maxonfjvipon** [comment]
@Graur Hey, i'm working on `pow` method for infinities and I have a question. We talk about `negative-infinity.pow n`.  
If `n > 0` and `n` is odd - result is `negative-infinity`, if `n` is even - `positive-infinity`. All this infinities stuff is about float pointing numbers, so I assume that `n` is float. But how can I check if this `n` is even or odd? `float` object doesn't have `mod` method, `int` has. But if I do `n.as-int.mod 2` - what will I get for n = 2.2? I think it'll be 0 and that says that 2 it's even but 2.2 - not.  
Any suggestions?

**Comment by maxonfjvipon** [comment]
@rultor merge

**Pull request merged**